content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
+prototypes, which act as _exemplars_ for other objects. Creating a new object
+from a prototype is as simple as _cloning_ it and setting its parent pointer to
+the prototype. Changing the common behavior of all clones is as simple as
+changing the behavior of a prototype. Since method and field lookups traverse
+the parent hierarchy, clones can also override methods and fields of their
+prototypes.
+
+A naive layout scheme for objects would copy all fields from a prototype and
+end up wasting a lot of space describing potentially shared behaviors. The Self
+compiler minimizes space usage of _clones_ derived from the same prototype by
+using _clone families_. A cloned object only stores its modifiable in its
+object and points to the clone family for its prototype. If the instance ever
+overrides one of its methods, the creates a new clone family to preserve
+semantics and propagate behavior changes to all clones of a prototype.
+
+<img src="without-maps.png" alt="drawing" width="300"/>
+<img src="with-maps.png" alt="drawing" width="300"/>
+
+### Customized Compilation
+
+The rampant dynamism in Self programs forces compilers to conservatively
+generate slow code. For example `x < y` is a method call on the object `x` and
+requires knowing the precise type of `x` to generate optimized code for `<`.
+The paper mentions that contemporary Smalltalk-80 compilers restricted the
+customization of primitive methods and control structures to allow for
+specialized code generation. Instead of imposing these restrictions, the Self
+compiler will generate specialized code for every _receiver_ object at runtime.
+This means that at the first invocation of `<` on an integer, the compiler will
+specialize `<` with check to see if the type matches to an `integer` and
+generate a single compare instruction when this is case. When the type of
+receiver has not been encountered before, Self will default to a method send
+to the object.
+
+### Message Inlining and Splitting
+
+Inlining remains one of the most crucial optimization in modern compilers,
+enabling other optimizations to become more effective. However, inlining
+becomes impossible in presence of dynamic method calls if the type of the
+receiver object is unknown. This is an even bigger problem for Self since
+most of program is a sequence of dynamic calls.
+
+The Self compiler performs two optimizations, message _inlining_ and
+_splitting_ to enable efficient execution. Message _inlining_ works in a similar
+fashion to customized compilation -- If the type of a receiver is know, either
+at initial compilation through a dataflow analysis or at runtime, inline the
+method body at the call location. When type information for an object is lost
+due to control flow splits, _message splitting_ generates specialized code for
+possible receiver types and guards them using type tests. If the type of
+receiver matches, fast code can be executed.
```

Comment:
To drive this point home, maybe it would be useful to include some example code to illustrate what the program looks like after these optimizations are applied. For example, you could show several imaginary assembly- or C-like listings depicting these versions of a program `x < y`:

- The obvious version, where the code looks up the string `lessThan` in the method dictionary for the object `x` and jumps to the corresponding code with argument `y`.
- The "customized compilation" version, with a type guard and a direct `call` to the `integer.lessThan` function with `y` as the argument.
- The "inlined" version, which still has the type guard but just uses a `lt` machine instruction or whatever to do the comparison—no call necessary.

For bonus points, you could even show imaginary code for the "split" version.